src/muya/lib/contentState/paragraphCtrl.js

```diff
@@ -834,6 +864,131 @@ const paragraphCtrl = ContentState => {
 
     return this.selectAllContent()
   }
+
+  // Test whether the paragraph transformation is valid.
+  ContentState.prototype.isAllowedTransformation = function (block, toType, isMultilineSelection) {
+    const fromType = this.getTypeFromBlock(block)
+    if (toType === 'front-matter') {
+      // Front matter block is added at the beginning.
+      return true
+    } else if (!fromType) {
+      return false
+    } else if (isMultilineSelection && /heading|table/.test(toType)) {
+      return false
+    } else if (fromType === toType || toType === 'reset-to-paragraph') {
+      // Convert back to paragraph.
+      return true
+    }
+
+    switch (fromType) {
+      case 'paragraph':
+        return !/hr|table/.test(toType)
+      case 'heading 1':
+      case 'heading 2':
+      case 'heading 3':
+      case 'heading 4':
+      case 'heading 5':
+      case 'heading 6':
+        return /paragraph|heading/.test(toType)
+      case 'ul-bulle':
+      case 'ul-task':
+      case 'ol-order':
+        return /ul|ol|loose-list-item/.test(toType)
+      default:
+        // Tables and all code blocks are not allowed.
+        return false
+    }
+  }
+
+  // Translate block type into internal name.
+  ContentState.prototype.getTypeFromBlock = function (block) {
```

Comment:
The parameter `block` will always be `span` block? so can we use `getAnchor` method to get the `block type`?

Reply:
> The parameter block will always be span block?

I don't know but I guess most of the time and I also fixed diagram, list and block quotes.

Reply:
It will always be type `span` block, and you can use `getAnchor` to markdown syntax block type, like table, diagram etc, is it useful for your case?

Reply:
`getAnchor` will help with with list items and quote blocks but return the same as I can read from block.

> It will always be type span block,

At least we have support for it. if we need it sometime.